ios emulation-based analysis

su-vikas opened the pull request:
For issue #1373

Added content for iOS emulation-based analysis section.

sushi2k commented:
Hi @su-vikas. Thanks for your PR. The iOS Simulator is already described in the iOS basic security testing chapter https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06b-Basic-Security-Testing.md#testing-on-the-ios-simulator.

Therefore your content fits much better into the iOS basic chapter. You could add your introduction to the section to make it also clearer what emulator/simulator means:

```
Emulation is imitation of a certain computer platform or program on another platform or program. The software or hardware performing this imitation is called an _emulator_. Emulator provides a much cheaper alternatives to an actual device, where a user can manipulate it without worrying about damaging the device. As discussed in "[Emulation-based analysis](0x05c-reverse-engineering-and-tampering#emulation-based-analysis "Emulation-based analysis")" section, there are multiple emulators available for Android, but unlike Android, for iOS there are practically no viable emulators available.
```

For Corellium you should also mention it in the basic testing chapter of iOS as part of the simulator paragraph, to be consistent with Android https://github.com/OWASP/owasp-mstg/blob/fc7d7bb1e534e6f27f0d581b0a08bee7cef3871a/Document/0x05b-Basic-Security_Testing.md#testing-on-an-emulator

Let's just make a very short description of Corellium as done in Android and you could mention that the product is an enterprise SaaS and there are no trial licenses available. 
We just need to be careful as Corellium is being sued and it's unclear what is happening with them.

sushi2k commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+##### Corellium
+
+Corellium is a commercial tool which offers virtual iOS devices - a virtual iPhone running actual iOS firmware. The product is very expensive (IDA-Pro license will feel very cheap) and there are no trial licenses available. This product is discussed here as it is the only publicly available iOS emulator ever.
```
See main comment

su-vikas commented:
Hi @sushi2k. Thanks for the comments. I will move the content to basic security testing chapter.
Then practically there is no need of this sub-section on emulator in 6c? Do you agree?

cpholguera commented:
I'd leave the section just to respect the structure that mirrors both chapters. You could move most of the content where Sven suggests and I'd leave here 1 or 2 sentences referring to that section in 6b and again, why it's not suitable for the testing we're doing in this whole chapter :)

Background: the outlines of the 5c / 6c should be almost the same, i.e. mirror each other as much as possible as they were redesigned to give the reader a feeling of "finding the same information" everywhere. (Even when it's to say that something is not possible but on the other platform it is)

su-vikas commented:
@cpholguera thanks for clarification. I will update and push the changes.

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+Apple provides a simulator app within Xcode which provides an user interface for iPhone, iPad or Apple Watch. As per Apple documentation, "simulator allows you to rapidly prototype and test builds of your app during the development process." Although the iOS simulator provides a similar look and feel as a real iOS device and also allows to run a debug application, but actually it is not an emulat ...
+
+An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment. QEMU based emulators for Android take into consideration the RAM, CPU, battery performance etc (hardware components) while running an application, but in an iOS simulator these hardware component behaviour is not taken into consideration at all ...
```
I'd also put this in 4c. It also gives a nice general overview.

I've added part of the previous paragraph here.

```suggestion
The difference between a simulator and an emulator often causes confusion and leads to use of the two terms interchangeably, but in reality they are different, specially for the iOS use case. An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment.

QEMU based emulators for Android take into consideration the RAM, CPU, battery performance etc (hardware components) while running an application, but in an iOS simulator this hardware component behaviour is not taken into consideration at all. The iOS simulator even lacks the implementation of the iOS kernel, as a result if an application is using syscalls it cannot be executed in this simulator.

In simple words, an emulator is a much closer imitation of the targeted platform, while a simulator mimics only a part of it. 
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment. QEMU based emulators for Android take into consideration the RAM, CPU, battery performance etc (hardware components) while running an application, but in an iOS simulator these hardware component behaviour is not taken into consideration at all ...
+
+iOS simulator only runs on an x86 platform, and as a consequence applications having code compiled in x86 instruction set only can be executed on it. While developing and debugging an application, Xcode toolchain do generate x86 code, but for a release build the code is compiled into ARM instruction set only, i.e no x86 code is generated. Due to the absence of x86 code in the release build, an ap ...
```
I'd leave this here but condensate it in shorter sentences, what do you think of this?

```suggestion
While developing and debugging an application, the Xcode toolchain generates x86 code, which can be executed in the iOS simulator. However, for a release build, only ARM code is generated (incompatible with the iOS simulator). That's why applications downloaded from the Apple App Store cannot be used for any kind of application analysis on the iOS simulator.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+iOS simulator only runs on an x86 platform, and as a consequence applications having code compiled in x86 instruction set only can be executed on it. While developing and debugging an application, Xcode toolchain do generate x86 code, but for a release build the code is compiled into ARM instruction set only, i.e no x86 code is generated. Due to the absence of x86 code in the release build, an ap ...
+
+In summary, iOS simulator is not an emulator and an application from Apple App Store cannot be executed as it does not have the x86 compiled code.
```
I'd remove this completely as, after the other changes the section will be very straightforward and therefore doesn't need a summary ;)
```suggestion
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+##### iOS Simulator
+
+Apple provides a simulator app within Xcode which provides an user interface for iPhone, iPad or Apple Watch. As per Apple documentation, "simulator allows you to rapidly prototype and test builds of your app during the development process." Although the iOS simulator provides a similar look and feel as a real iOS device and also allows to run a debug application, but actually it is not an emulat ...
```
I'd condensate this and move the last part to the next paragraph:

```suggestion
Apple provides a simulator app within Xcode which provides a _real iOS device looking_ user interface for iPhone, iPad or Apple Watch. It allows you to rapidly prototype and test debug builds of your applications during the development process, but actually **it is not an emulator**.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+##### Corellium
+
+Corellium is a commercial tool which offers virtual iOS devices - a virtual iPhone running actual iOS firmware. The product is very expensive (IDA-Pro license will feel very cheap) and there are no trial licenses available. This product is discussed here as it is the only publicly available iOS emulator ever.
```
I've integrated a part of the next paragraph here and made clear that it is an emulator (at the beginning already, so no more explanation would be required).

The basic message for this and next paragraph: "it's an actual emulator that can be used to analyze release apps but we won't go into detail as it is commercial".

```suggestion
Corellium is a commercial tool which offers virtual iOS devices running actual iOS firmware, being the only publicly available iOS emulator ever. Since it is a proprietary product, not much information is available about the implementation. Corellium has an elevated price and there are no trial or community licenses available, therefore we won't go into much detail regarding its use.
```

su-vikas replied:
can we use other word than "elevated price". I have never seen elevated used in such context? do you agree?

cpholguera replied:
I agree and adapted the suggestion. What do you think?

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -436,6 +436,27 @@ Next, navigate to a new website in Safari. You should see traced function calls
+
+Corellium is a commercial tool which offers virtual iOS devices - a virtual iPhone running actual iOS firmware. The product is very expensive (IDA-Pro license will feel very cheap) and there are no trial licenses available. This product is discussed here as it is the only publicly available iOS emulator ever.
+
+Since it is a proprietary product, not much information is available about the implementation, but it is safe to assume that some kind of emulation is used behind the scenes. For an end user it is an emulator, with which multiple instances of a device can be launched, and the ability to take a snapshot of the device state and restore it, if needed. The product is quite straightforward to use and  ...
```
```suggestion
Corellium allows you to launch multiple instances of a device (jailbroken or not) which are accessible as local devices (with a simple VPN configuration). It has the ability to take and restore snapshots of the device state, and also offers a convenient web-based shell to the device. Finally and most importantly, due to its "emulator" nature, you can execute applications downloaded from the Apple App Store, enabling any kind of application analysis as you know it from real iOS (jailbroken) devices.
```

su-vikas replied:
nice!

su-vikas commented:
@sushi2k @cpholguera I have updated the content as per your comments. Please have a re-look. I hope I moved the content correctly to various sections.

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -438,23 +438,24 @@ Next, navigate to a new website in Safari. You should see traced function calls
-Apple provides a simulator app within Xcode which provides an user interface for iPhone, iPad or Apple Watch. As per Apple documentation, "simulator allows you to rapidly prototype and test builds of your app during the development process." Although the iOS simulator provides a similar look and feel as a real iOS device and also allows to run a debug application, but actually it is not an emulat ...
+Apple provides a simulator app within Xcode which provides a _real iOS device looking_ user interface for iPhone, iPad or Apple Watch. It allows you to rapidly prototype and test debug builds of your applications during the development process, but actually **it is not an emulator**.
+
+The difference between a simulator and an emulator often causes confusion and leads to use of the two terms interchangeably, but in reality they are different, specially for the iOS use case. An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment.
```
I think I'd actually take from here 445 until 449 and put this explanation in 0x4c as well. It's very nice to have it in one place talking about both platforms and to keep this section short. That would leave us here with only 2 nice paragraphs going straight to the point ;)

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -438,23 +438,24 @@ Next, navigate to a new website in Safari. You should see traced function calls
 ##### iOS Simulator
 
-Apple provides a simulator app within Xcode which provides an user interface for iPhone, iPad or Apple Watch. As per Apple documentation, "simulator allows you to rapidly prototype and test builds of your app during the development process." Although the iOS simulator provides a similar look and feel as a real iOS device and also allows to run a debug application, but actually it is not an emulat ...
+Apple provides a simulator app within Xcode which provides a _real iOS device looking_ user interface for iPhone, iPad or Apple Watch. It allows you to rapidly prototype and test debug builds of your applications during the development process, but actually **it is not an emulator**.
```
After doing what's mentioned below I'd end this paragraph with a reference to 0x4c section. Something like: `See section "[Emulation-based Dynamic Analysis](...)" for more details.`

cpholguera commented on Document/0x04c-Tampering-and-Reverse-Engineering.md:
```diff
@@ -184,7 +184,10 @@ The [Valgrind README](http://valgrind.org/docs/manual/dist.readme-android.html "
 #### Emulation-based Dynamic Analysis
 
-Running an app in the emulator gives you powerful ways to monitor and manipulate its environment. For some reverse engineering tasks, especially those that require low-level instruction tracing, emulation is the best (or only) choice. Unfortunately, this type of analysis is only viable for Android, because no emulator exists for iOS (the iOS simulator is not an emulator, and apps compiled for an  ...
+Emulation is an imitation of a certain computer platform or program on another platform or program. The software or hardware performing this imitation is called an _emulator_. Emulator provides a much cheaper alternatives to an actual device, where an user can manipulate it without worrying about damaging the device. There are multiple emulators available for Android, but unlike Android, for iOS  ...
```
I've enhanced the first sentence. Does it still express the idea as intended?

```suggestion
Emulation is an imitation of a certain computer platform or program being executed in different platform or within another program. The software or hardware performing this imitation is called an _emulator_. Emulators provide a much cheaper alternative to an actual device, where a user can manipulate it without worrying about damaging the device. There are multiple emulators available for Android, but for iOS there are practically no viable emulators available.
```

su-vikas commented:
@cpholguera fixed as per your comments.
